File: frontend/src/services/holdingService.js
import api from './api';

export const getHoldings = (portfolioId) => api.get(`/portfolio/${portfolioId}/holdings`);
export const addHolding = (portfolioId, data) => api.post(`/portfolio/${portfolioId}/holdings`, data);
export const updateHolding = (portfolioId, holdingId, data) => api.put(`/portfolio/${portfolioId}/holdings/${holdingId}`, data);
export const deleteHolding = (portfolioId, holdingId) => api.delete(`/portfolio/${portfolioId}/holdings/${holdingId}`);
export const getTransactions = (portfolioId, holdingId) => api.get(`/portfolio/${portfolioId}/holdings/${holdingId}/transactions`);
export const addTransaction = (portfolioId, holdingId, data) => api.post(`/portfolio/${portfolioId}/holdings/${holdingId}/transactions`, data);

File: backend/src/main/java/com/investmentresearch/service/HoldingService.java
package com.investmentresearch.service;

import com.investmentresearch.dto.*;

import java.util.List;
import java.util.UUID;

public interface HoldingService {

    List<HoldingResponse> getHoldings(UUID portfolioId, String email);

    HoldingResponse addHolding(UUID portfolioId, HoldingRequest request, String email);

    HoldingResponse updateHolding(UUID holdingId, UpdateHoldingRequest request, String email);

    void deleteHolding(UUID holdingId, String email);

    List<TransactionResponse> getTransactions(UUID holdingId);

    TransactionResponse addTransaction(UUID holdingId, TransactionRequest request, String email);
}
